### DISCUSSION TITLE
Exception when use requires

### DISCUSSION THREAD

**Discussion by xidianwlc** [fastapi/fastapi#7895] [Questions] [answered]
### exception when use requires

starlette/authentication.py  line 60

        elif asyncio.iscoroutinefunction(func):
            # Handle async request/response functions.
            @functools.wraps(func)
            async def async_wrapper(
                *args: typing.Any, **kwargs: typing.Any
            ) -> Response:
                request = kwargs.get("request", args[idx])   #  this place Exception when the idx=0 but args = ()
                assert isinstance(request, Request)

### The Exception context

[ERROR 2020-01-13 20:20:33,617 httptools_impl.py 388] Exception in ASGI application
Traceback (most recent call last):
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/uvicorn/protocols/http/httptools_impl.py", line 385, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/fastapi/applications.py", line 140, in __call__
    await super().__call__(scope, receive, send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/applications.py", line 134, in __call__
    await self.error_middleware(scope, receive, send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/middleware/errors.py", line 178, in __call__
    raise exc from None
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/middleware/errors.py", line 156, in __call__
    await self.app(scope, receive, _send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/middleware/authentication.py", line 48, in __call__
    await self.app(scope, receive, send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/middleware/base.py", line 25, in __call__
    response = await self.dispatch_func(request, self.call_next)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette_exporter/middleware.py", line 40, in dispatch
    raise e
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette_exporter/middleware.py", line 37, in dispatch
    response = await call_next(request)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/middleware/base.py", line 45, in call_next
    task.result()
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/middleware/base.py", line 38, in coro
    await self.app(scope, receive, send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/exceptions.py", line 73, in __call__
    raise exc from None
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/exceptions.py", line 62, in __call__
    await self.app(scope, receive, sender)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/routing.py", line 590, in __call__
    await route(scope, receive, send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/routing.py", line 208, in __call__
    await self.app(scope, receive, send)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/routing.py", line 41, in app
    response = await func(request)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/fastapi/routing.py", line 127, in app
    raw_response = await dependant.call(**values)
  File "/data/server/pyenv/py3.7.4/lib/python3.7/site-packages/starlette/authentication.py", line 62, in async_wrapper
    request = kwargs.get("request", args[idx])
IndexError: tuple index out of range

### To produce

from fastapi import APIRouter, Query, Depends, Header, HTTPException
from starlette.authentication import requires
from starlette.requests import Request

from zyservice.user.auth import AuthAPI, LoginResponse
from zycommon.helper.auth import get_current_user, JWTUser

router = APIRouter()

@router.post("/login", response_model=LoginResponse)
async def login(
    *,
    username: str = Query(..., max_length=100, description="user name"),
    password: str = Query(
        ..., min_length=8, max_length=100, description="user password"
    ),
):
    """ login
    """
    return await AuthAPI().login(username, password)

@router.get("/info")
@requires("authenticated")
async def user_info(*, request: Request, user: JWTUser = Depends(get_current_user)):
    """ get user info
    """
    return {"username": user.username}

### Environment

- OS: [e.g. Linux ]
- FastAPI Version [0.46.0, 0.45.0]

**Comment by dmontagu** [comment]
Could you describe what you are trying to achieve?

------

This seems more likely to be a starlette issue than a fastapi issue; it's probably worth posting there instead.

-------

The formatting of your post makes it very hard to follow what's going on -- could you enter the code like this:

\`\`\`
code goes here
\`\`\`

That should ensure it gets formatted to look like this:
```
code goes here
```
which is much easier to read.

**Comment by xidianwlc** [comment, marked as answer]
This is starlette's bug, I will post it there.

**Comment by tiangolo** [comment]
Thanks for the help here @dmontagu ! :cake: :bowing_man: 

And thanks @xidianwlc  for reporting back and closing the issue :+1:
